skbio/tree/_dists.py

```diff
@@ -0,0 +1,395 @@
+
+    Parameters
+    ----------
+    trees : list of TreeNode
+        Input trees.
+    ids : list of str, optional
+        Unique identifiers of input trees. If omitted, will use incremental integers
+        "0", "1", "2",...
+    pairwise : bool, optional
+        Calculate the distance between each pair of trees based on taxa shared across
+        all trees (False, default), or shared between the current pair of trees (True).
+    proportion : bool, optional
+        Whether to return the RF distance as count (False, default) or proportion
+        (True).
+    rooted : bool, optional
+        Whether to consider the trees as unrooted (False, default) or rooted (True).
+
+    Returns
+    -------
+    DistanceMatrix
+        Matrix of the Robinson-Foulds distances.
+
+    See Also
+    --------
+    TreeNode.compare_rfd
+
+    Notes
+    -----
+    The Robinson-Foulds (RF) distance [1]_, a.k.a. symmetric difference, is the number
+    of bipartitions differing between two trees.
+
+    This function is equivalent to :meth:`TreeNode.compare_rfd` for two trees. Refer
+    to the latter for details about the metric and its parameters. However, the current
+    function extends the operation to an arbitrary number of trees and returns a
+    distance matrix for them.
+
+    This function is optimized for calculation based on taxa shared across all trees.
+    One can instead set ``pairwise`` to True to calculate based on taxa shared between
+    each pair of trees, which is however less efficient since bipartitions need to be
+    re-inferred during each comparison.
+
+    References
+    ----------
+    .. [1] Robinson, D. F., & Foulds, L. R. (1981). Comparison of phylogenetic
+        trees. Mathematical biosciences, 53(1-2), 131-147.
+
+    Examples
+    --------
+    >>> trees = [TreeNode.read([x]) for x in (
+    ...     "(((a,b),c),d,e);",
+    ...     "((a,(b,c)),d,e);",
+    ...     "((a,b),(c,d),e);",
+    ...     "(a,b,(c,(d,e)));",
+    ... )]
+    >>> dm = rf_dists(trees, ids=list("ABCD"))
+    >>> print(dm)
+    4x4 distance matrix
+    IDs:
+    'A', 'B', 'C', 'D'
+    Data:
+    [[ 0.  2.  2.  0.]
+     [ 2.  0.  4.  2.]
+     [ 2.  4.  0.  2.]
+     [ 0.  2.  2.  0.]]
+
+    """
+    _check_ids(trees, ids)
+    if pairwise:
+        metric = partial(TreeNode.compare_rfd, proportion=proportion, rooted=rooted)
+        return DistanceMatrix.from_iterable(
+            trees, metric=metric, keys=ids, validate=False
+        )
+
+    method = "subsets" if rooted else "biparts"
+    withins = _withins(trees)
+    sets = [getattr(x, method)(within=y) for x, y in zip(trees, withins)]
+    lens = [len(s) for s in sets]
+
+    result = np.zeros((n_trees := len(trees), n_trees))
+    for i, j in combinations(range(n_trees), 2):
+        rf = len(sets[i] ^ sets[j])
+        if proportion:
+            total = lens[i] + lens[j]
+            rf = rf / total if total else 1.0
+        result[i, j] = result[j, i] = rf
```

Comment:
It looks like this may be a re-implementation of RFD? I understand that this may be faster than to filter the trees and use `TreeNode.compare_rfd`, and suspect that's why it was done. However, it comes at the expense of maintenance burden, the potential for differing results based on the exact implementation used, and may violate the DRY principle. 

If the existing `compare_rfd` method is used, then there is a single point in the codebase to maintain for RFD, and a single block of code to consider optimizing in the future.

Reply:
Good point. I am still in the process of refining the code. Ideally, each algorithm is only implemented once. There are some discrepancies between two tree RF and multi-tree RF, because the former also takes care of some non-RF metrics. But I might be able to overcome this.

Reply:
I wonder whether a single implementation of RF that accepts `Iterable[set | frozen set]` (i.e., agnostic to `TreeNode`, etc), possibly an implementation specific for wRF, should be provided. And then this code, as well as the respective `TreeNode.compare_*` methods call that single stable implementation?

Reply:
I have implemented your suggestion. Specifically, in `_compare.py` there is a driver function `_topo_dists` which calls `_setdiff` or `_mapdiff` depending on whether trees are rooted.